add map_name commands

Berthalamew opened the pull request:
- added other commands to set local player counts, difficulty, game variant, and game mode

Berthalamew committed:
add map_name commands

- added other commands to set local player counts, difficulty, game variant, and game mode

Berthalamew committed:
fix incorrect typename

Berthalamew committed:
fix compiler warnings and crash in release mode

Berthalamew committed:
misc changes

- add console logging for improper values
- add k_campaign_difficulty_levels_count define
- rename game_set_difficulty to main_game_launch_set_difficulty

Berthalamew committed:
add cseries macros for range checks

thanks twister

nukeulater commented on xlive/Blam/Engine/main/main_game.cpp:
```diff
@@ -0,0 +1,213 @@
+#include "Util/Hooks/Hook.h"
+
+s_game_options g_main_game_launch_options;
+DWORD g_main_game_launch_user_count;
```
Nit int

Berthalamew replied:
the two globals are initialized now.

nukeulater replied:
Meant you should make the variable int instead.

nukeulater commented on xlive/Blam/Engine/main/main_game.cpp:
```diff
@@ -0,0 +1,213 @@
+void main_game_launch_set_difficulty(short difficulty)
+{
+    if (!VALID_INDEX(difficulty, k_campaign_difficulty_levels_count))
+    {
```
Just return here if validation fails and remove the else statement. Also do it for the other similar cases

nukeulater commented on xlive/Blam/Engine/main/main_game.cpp:
```diff
@@ -0,0 +1,213 @@
+    {
+    case _game_mode_campaign:
+    {
+        if (g_main_game_launch_options.difficulty >= 0)
```
Use the range macros you just added.

Berthalamew replied:
changed

nukeulater commented on xlive/H2MOD/GUI/imgui_integration/Console/CommandsUtil.h:
```diff
@@ -193,25 +193,31 @@ class CircularStringBuffer
 		va_start(valist, fmt);
 		int buffer_size_needed = _vsnprintf(NULL, 0, fmt, valist) + 1;
-		if (buffer_size_needed < m_line_buf_size)
+		if (buffer_size_needed != -1)
```
Will be -1 only when the result is -2. Practically never, because that function only returns - 1 if the buffer is insufficient.

Berthalamew replied:
changed so it makes sure the value is positive, need this since we cast it to a size_t type afterwards and don't want it to cause undefined behaviour.

nukeulater commented on xlive/H2MOD/GUI/imgui_integration/Console/CommandCollection.cpp:
```diff
@@ -769,3 +776,56 @@ void CommandCollection::DeleteObject(datum objectDatumIdx)
+int CommandCollection::map_name(const std::vector<std::string>& tokens, ConsoleCommandCtxData cbData)
+{
+	std::string name(tokens[1].begin(), tokens[1].end());
+	main_game_launch(name.c_str());
```
Why not just access the token directly, in the argument list, instead of creating another string?

Berthalamew replied:
Stack corruption happens when you try and do that on release builds

nukeulater replied:
How could that cause stack corruption?

Berthalamew replied:
No clue, something to do with cleanup after the map_name function is called from the console

Berthalamew committed:
add requested changes and other misc changes

- fix buffer_size_needed if statements in CommandsUtil
- initialize g_main_game_launch_user_count to 1
- make a function for setting the details for every game mode rather than including it within the switch case itself in main_game_launch_setup_game_mode_details
- use range check macros in game mode functions

Berthalamew committed:
initialize g_main_game_launch_options to 0
